pkg/handler/forwarder/s3http/internal/decoders/csv.go

```diff
@@ -74,20 +75,35 @@ func (dec *CSVDecoder) Decode(v any) error {
 		return fmt.Errorf("failed to decode record: %w", err)
 	}
 
-	var buf bytes.Buffer
+	buf := bytes.NewBuffer(make([]byte, 0, dec.maxSize))
 
 	buf.WriteString(`{`)
 	for i, colName := range dec.header {
 		if i < len(record) && record[i] != "" {
 			if buf.Len() != 1 {
 				buf.WriteString(`,`)
 			}
-			buf.WriteString(colName + `:` + strconv.Quote(record[i]))
+
+			buf.WriteString(colName + `:`)
+
+			// it is cheaper to verify if naive quoting is enough
+			if value := []byte(`"` + record[i] + `"`); json.Valid(value) {
+				buf.Write(value)
+			} else {
+				buf.WriteString(strconv.Quote(record[i]))
+			}
 		}
 	}
 	buf.WriteString(`}`)
 
-	if err := json.Unmarshal(buf.Bytes(), v); err != nil {
+	if buf.Len() > dec.maxSize {
+		dec.maxSize = buf.Len()
+	}
+
+	// avoid unmarshalling if possible
+	if r, ok := v.(*json.RawMessage); ok {
```

Comment:
no additional tests for this new branch?

Reply:
This is actually the only case we test today, since both the tests and the runner use json.RawMessage. Arguably we could add a testcase for the unused path, but that was tested prior so we do know it works. We would be ensuring that it continues to work going forward. I've update the test to verify this slow path. The test results differ due to key reordering (when we go through `json.Unmarshal`, keys are ordered alphabetically)